### ISSUE TITLE
Cannot set Foreign Key column name by Fluent API

### ISSUE THREAD

**Issue by cdie** [aspnet/EntityFramework#6880]
I'm trying to give my Foreign Key column the name I want (not talking about the constraint name).

If I set my foreignKey with the builder.Property, 

```
var b = builder.Property(column.PropertyType, column.PropertyName);
b.ForSqlServerHasName("MyFKColumnName");
```

when I create relationship, I have an error on migration : 
"The navigation property 'LettreClePrincipale' cannot be added to the entity type 'LettreCleAssociation' because a property with the same name already exists on entity type 'LettreCleAssociation'."

If I don't set my foreignKey with the builder.Property, this error is gone, the key is correctly created, but I cannot set name during the process : 

```

builder.HasMany(fks.DistantObjectType.FullName, fks.CurrentPropertyName)
.WithOne().IsRequired(fks.Required)
.OnDelete(fks.DeleteCascade ? Microsoft.EntityFrameworkCore.Metadata.DeleteBehavior.Cascade : Microsoft.EntityFrameworkCore.Metadata.DeleteBehavior)
.HasForeignKey(keyStorageOfProp.Name).???
```

I don't have ForSqlServerHasName or ForSqliteHasName extensions method here. How can I proceed to achieve this ?

**Comment by rowanmiller** [comment]
Can you share a full code listing (or project) that we can run to debug the issue? There isn't enough info here for us to work out what is going on.

**Comment by cdie** [comment]
Considering this architecture

```
class LTC
        {
            public int TestA { get; set; }
            public string TestB { get; set; }

            public Guid Id { get; set; }

            protected ICollection<LTCLinks> _links { get; set; }

            public IReadOnlyList<LTCLinks> Links
            {
                get
                {
                    return _links.ToList().AsReadOnly();
                }
            }
        }

        class LTCLinks
        {
            public LTC Link1 { get; set; }
            public LTC Link2 { get; set; }

            public string Data { get; set; }
        }
```

with this configuration 

```
                var ltcBuilder = modelBuilder.Entity(typeof(LTC));
                ltcBuilder.Property(typeof(int), "TestA").IsRequired();
                ltcBuilder.Property(typeof(string), "TestB").IsRequired();
                ltcBuilder.HasKey("Id");
                ltcBuilder.Ignore("_links");


                var ltcLinkBuilder = modelBuilder.Entity(typeof(LTCLinks));
                ltcBuilder.Property(typeof(string), "Data").IsRequired();

                ltcLinkBuilder.HasOne(typeof(LTC).FullName, "Link1").WithMany().IsRequired().OnDelete(Microsoft.EntityFrameworkCore.Metadata.DeleteBehavior.Cascade);
                ltcLinkBuilder.HasOne(typeof(LTC).FullName, "Link2").WithMany().IsRequired(false).OnDelete(Microsoft.EntityFrameworkCore.Metadata.DeleteBehavior.Restrict);

                ltcLinkBuilder.HasKey("Data", "Link1ID", "Link2ID");
```

The two foreign keys for LTCLinks are created as shadow properties by EF. How can I define their name in fluent API ?
I've found a workaround for my case by creating explicit foreign key and giving they a name by using using the entityBuilder.Property method. I think it's the good way, but I want to have a workaround for my DAL if the final dev has not created the foreign key.

**Comment by divega** [comment]
> I've found a workaround for my case by creating explicit foreign key and giving they a name by using using the entityBuilder.Property method.

@cdie I am not sure if by "creating explicit foreign key" you are referring to defining a property explicitly with a call like `entityBuilder.Property<Guid>("Link1Id")` or if you are adding actual C# properties to the `LTCLinks` type. Either approach should work, but I would also recommend adding a call to `HasForeignKey()` indicating what property to use for each relationship to avoid any ambiguity.

I am also curious about some things you are doing in your code, e.g. why not use the generic versions of the model building methods and lambda expressions to identify properties? It can lead to easier to read code.

**Comment by cdie** [comment]
@divega when I said "creating explicit foreign key", I was talking about adding 2 Guids C# properties for defining foreign keys for `Link1 `& `Link2 `object. In fact, these properties are not mandatory for EF to work, as I said in my first post, EF create shadow properties for them if we don't add them.
If the final developper in our company forget to create those 2 Guids properties, I don't know how to set the foreign key name. And yes, indeed, in my workaround, because I created these two properties, I can use the `HasForeignKey` method. However, same question, how can I do that if it's EF that create shadow properties for me ? 
This is more curiousity, because I'll try to force our devs to create foreign key properties by throwing exception during migration creation.

About the using of generic versions, the code above was just a little example of how when proceed, because our architecture was previously used without EF, so we've defined our own attributes to create migrations automaticly by reflection. So I created an example and adapted it to try to retranscript at best what we're doing by reflection for our own business objects. I wanna be sure that it's (or not) possible to fix foreign key name with non generics methods, so this is why I designed such example.

**Comment by smitpatel** [comment]
How to name the FK column
- If you are having a property in CLR type which is being used as FK property then easiest way to do is to use `Property().HasColumnName()`
e.g. `modelBuilder.Entity<LTCLinks>().Property(e => e.Link1Id).HasColumnName("MyName");`

- If you are not having a property in CLR type then you would have a shadow property for FK column. Now you can either name your shadow property as the column name you want or leave the name of FK by convention and explicitly name it to something different. Keep in mind that while the conventional FK name remains the same but the name depends on the model due to clash in the name between multiple relationships. So it may not possible for you to compute the exact name for every model easily and it is not advisable to go that path. Easiest way is the former - Name your shadow FK the name you want in database. This way you don't have to have explicit CLR property still you can control the name.
e.g. `builder.Entity<LTC>().HasMany<LTCLinks>().WithOne(l => l.Link1).HasForeignKey("MyName");`

**Comment by smitpatel** [comment]
@divega - Should we close this issue now?
The issue was created before 1.1 release. It seems that `.Property` method was used on navigation property to configure the name but then later used in relationship API hence the exception thrown. We have improved a lot of more checks in metadata API to avoid such incorrect model building.
Also, I responded above how to change the name of FK property if user wants to.
